Add reference for asm-goto

nbdd0121 opened the pull request:
The asm-goto-with-outputs is still unstable, so in the reference it's still mentioned as forbidden by a check rule.

Stabilization PR: https://github.com/rust-lang/rust/pull/133870

nbdd0121 committed:
Add reference for asm-goto

The asm-goto-with-outputs is still unstable, so in the reference it's
still mentioned as forbidden by a check rule.

Amanieu commented on src/inline-assembly.md:
```diff
@@ -180,6 +180,13 @@ r[asm.operand-type.supported-operands.sym]
+  - The address of the block is substituted into the asm template string. The assembly block may jump to the substituted addresses.
+  - After execution of the block, the `asm!` expression returns.
+  - The type of the block must be unit or `!` (never).
+  - The block starts new safety context; despite the outer `unsafe` needed for `asm!`, you need extra unsafe to perform unsafe operations inside the block.
```
```suggestion
  - The block starts new safety context; despite the outer `unsafe` needed for `asm!`, you need an extra `unsafe` to perform unsafe operations inside the block.
```

Amanieu commented on src/inline-assembly.md:
```diff
@@ -553,6 +560,8 @@ r[asm.options.supported-options.noreturn]
 - `noreturn`: The `asm!` block never returns, and its return type is defined as `!` (never).
   Behavior is undefined if execution falls through past the end of the asm code.
   A `noreturn` asm block behaves just like a function which doesn't return; notably, local variables in scope are not dropped before it is invoked.
+- When labels are present, `noreturn` means the `asm!` block never fallthrough; the asm block may jump to one of the specified blocks.
```
```suggestion
- When labels are present, `noreturn` means the execution in the `asm!` block never falls through; the asm block may only exit by jumping to one of the specified blocks.
```

nbdd0121 commented on src/inline-assembly.md:
```diff
@@ -576,7 +585,10 @@ r[asm.options.checks.pure]
 
 r[asm.options.checks.noreturn]
-- It is a compile-time error to specify `noreturn` on an asm block with outputs.
+- It is a compile-time error to specify `noreturn` on an asm block with outputs and without labels.
```
This change is to make `r[asm.options.checks.label-with-outputs]` be the only thing that blocks the `asm-goto-with-outputs` feature.

traviscross committed:
Make some editorial adjustments

nbdd0121 commented:
> In reading this more carefully (especially the lines not changed), I find myself wanting to double check that we're being precise enough now by what we mean by "the `asm!` block". Previously, when speaking about what happened during execution of "the `asm!` block", that meant only the execution of the assembly code. But now, given that the `label` blocks are lexically within the `asm!` block and may be executed during the execution of the `asm!` block, it would be reasonable to read "the `asm!` block" as to include that code also.
> 
> So when we say, e.g., that...
> 
> > * `pure`: The `asm!` block has no side effects, must eventually return, and its outputs depend only on its direct inputs (i.e. the values themselves, not what they point to) or values read from memory (unless the `nomem` options is also set).

GCC manual says "Also note that an asm goto statement is always implicitly considered volatile.", although AFAIU it's possible to represent a asm-goto in LLVM without the `sideeffect` modifier. Clang generates LLVM IR differently for `asm goto` and `asm goto volatile`, however LLVM doesn't seem to handle it specially (e.g. function containing a "pure" asm-goto block isn't treated as pure and is not optimised out).

It's not clear what exactly is pure for an asm goto block, since the control flow change is also one of its output. We probably should just forbid pure + label from being used together -- however note that since `pure` require an output operand, this is currently forbidden without a stable `asm_goto_with_outputs` feature.

> ...or...
> 
> > * `nomem`: The `asm!` block does not read from or write to any memory accessible outside of the `asm!` block.
> 
> ...or...
> 
> > * `readonly`: The `asm!` block does not write to any memory accessible outside of the `asm!` block.
> 
> ...or even...
> 
> > * `nostack`: The `asm!` block does not push data to the stack...
> 
> ...do we mean for these to apply to the `label` blocks also? I wouldn't think so, but it's rather unclear in reading it.

These would just mean the assembly themselves.

nbdd0121 commented:
> I wouldn't think so, but it's rather unclear in reading it.

How about replacing "`asm!` block" with "assembly code"?

nbdd0121 committed:
Differentiate "assembly code" and "asm block".

The latter now only means the entire block, while the former means the
assembly code specified within the block (thus excluding `label` blocks).

Darksonn commented on src/inline-assembly.md:
```diff
@@ -354,6 +354,23 @@ assert_eq!(y, [3, 2, 0, 1]);
+  - The address of the block is substituted into the asm template string. The assembly code may jump to the substituted address.
+  - After execution of the block, the `asm!` expression returns.
+  - The type of the block must be unit or `!` (never).
+  - The block starts a new safety context; despite the outer `unsafe` block needed for `asm!`, unsafe operations within the `label` block must be wrapped in an inner `unsafe` block.
```
```suggestion
  - The block starts a new safety context: unsafe operations within the `label` block must be wrapped in an inner `unsafe` block, even though the entire `asm!` statement is already wrapped in `unsafe`.
```

Darksonn commented on src/inline-assembly.md:
```diff
@@ -1308,8 +1330,8 @@ r[asm.rules.arm64ec]
-- The requirement of restoring the stack pointer and non-output registers to their original value only applies when exiting an `asm!` block.
-  - This means that `asm!` blocks that never return (even if not marked `noreturn`) don't need to preserve these registers.
+- The requirement of restoring the stack pointer and non-output registers to their original value only applies when exiting the assembly code.
+  - This means that assembly code that never return (even if not marked `noreturn`) don't need to preserve these registers.
```
Most of the `asm!` block to assembly code changes need some grammar fixups. For example:
```suggestion
  - This means that assembly code that never returns (even if not marked `noreturn`) doesn't need to preserve these registers.
```

nbdd0121 committed:
Apply editorial suggestions from code review

Co-authored-by: Alice Ryhl <aliceryhl@google.com>

joshtriplett committed:
More distinctions between asm blocks and assembly code

Update more cases where the phrasing could potentially have been
interpreted as affecting a `label` block.

ehuss commented:
@joshtriplett Indicated they could give a review over the chapter to make sure everything still fits together.

joshtriplett committed:
Clarify an interaction between `noreturn` and `label`

joshtriplett commented:
I reviewed the entire chapter, and wrote https://github.com/nbdd0121/reference/pull/1 as a PR to this PR branch, since there were changes in areas outside the context of the patch, which github doesn't allow suggestions in.

I added some more disambiguations in cases of possible ambiguity between the entire asm! block (including label blocks) and the assembly code within it.

The second commit, which I've kept separate, attempts to clarify an interaction between noreturn and label blocks; however, for that commit I could use some confirmation on accuracy. The phrasing as previously written said that "A noreturn asm block behaves just like a function which doesn't return", but that isn't accurate in the presence of label blocks, since the asm block can return, from the label blocks. I changed it to apply to "A noreturn asm block with no label blocks". However, this assumes there are not more semantic connotations of noreturn here that do apply to asm blocks with label blocks.

With the first commit applied, and either the second commit (if accurate) or something like it (if another phrasing would be more accurate) applied, the chapter looks fully consistent to me.

traviscross committed:
Revise `asm_goto` changes

Let's make some clarifications of substance and editorial tweaks to
the changes for `asm_goto`.  We add an example, we change various
wording to be more clear -- in particular, we significantly revise the
section on `noreturn` to unify the statement of the rules -- and we
replace remaining uses of "asm code" with "assembly code" to match our
verbiage elsewhere.

The main clarification of substance is that for requirements that must
be upheld when falling through the assembly, such as he requirement to
restore the stack pointer, we must also note that the requirement must
be satisfied before jumping to any `label` blocks.
